core/src/banking_stage/worker.rs

```diff
@@ -0,0 +1,626 @@
+    },
+    crossbeam_channel::{select, Receiver, RecvError, SendError, Sender},
+    solana_poh::leader_bank_notifier::LeaderBankNotifier,
+    solana_runtime::bank::Bank,
+    std::{sync::Arc, time::Duration},
+    thiserror::Error,
+};
+
+#[derive(Debug, Error)]
+pub enum WorkerError {
+    #[error("Failed to receive work from scheduler: {0}")]
+    Recv(RecvError),
+    #[error("Failed to send finalized consume work to scheduler: {0}")]
+    ConsumedSend(SendError<FinishedConsumeWork>),
+    #[error("Failed to send finalized forward work to scheduler: {0}")]
+    ForwardedSend(SendError<FinishedForwardWork>),
+}
+
+impl From<RecvError> for WorkerError {
+    fn from(err: RecvError) -> Self {
+        Self::Recv(err)
+    }
+}
+
+impl From<SendError<FinishedConsumeWork>> for WorkerError {
+    fn from(err: SendError<FinishedConsumeWork>) -> Self {
+        Self::ConsumedSend(err)
+    }
+}
+
+impl From<SendError<FinishedForwardWork>> for WorkerError {
+    fn from(err: SendError<FinishedForwardWork>) -> Self {
+        Self::ForwardedSend(err)
+    }
+}
+
+pub(crate) struct Worker {
+    consume_receiver: Receiver<ConsumeWork>,
+    consumer: Consumer,
+    consumed_sender: Sender<FinishedConsumeWork>,
+
+    forward_receiver: Receiver<ForwardWork>,
+    forward_option: ForwardOption,
+    forwarder: Forwarder,
+    forwarded_sender: Sender<FinishedForwardWork>,
+
+    leader_bank_notifier: Arc<LeaderBankNotifier>,
+}
+
+impl Worker {
+    pub fn new(
+        consume_receiver: Receiver<ConsumeWork>,
+        consumer: Consumer,
+        consumed_sender: Sender<FinishedConsumeWork>,
+        forward_receiver: Receiver<ForwardWork>,
+        forward_option: ForwardOption,
+        forwarder: Forwarder,
+        forwarded_sender: Sender<FinishedForwardWork>,
+        leader_bank_notifier: Arc<LeaderBankNotifier>,
+    ) -> Self {
+        Self {
+            consume_receiver,
+            consumer,
+            consumed_sender,
+            forward_receiver,
+            forward_option,
+            forwarder,
+            forwarded_sender,
+            leader_bank_notifier,
+        }
+    }
+
+    pub fn run(self) -> Result<(), WorkerError> {
+        loop {
+            select! {
+                recv(self.consume_receiver) -> consume_work => {
+                    self.consume_loop(consume_work?)?;
+                },
+                recv(self.forward_receiver) -> forward_work => {
+                    self.forward_loop(forward_work?)?;
+                },
+            }
+        }
+    }
+
+    fn consume_loop(&self, consume_work: ConsumeWork) -> Result<(), WorkerError> {
+        let Some(mut bank) = self.get_consume_bank() else {
+            return self.retry_drain(consume_work);
+        };
+
+        for work in std::iter::once(consume_work).chain(self.consume_receiver.try_iter()) {
+            if bank.is_complete() {
+                if let Some(new_bank) = self.get_consume_bank() {
+                    bank = new_bank;
```

Comment:
race condition (extremely rate to happen in the wild, though): we need to check `new_bank` before entering `self.consume`.

i think `if bank.is_complete()` could be turned into `while !bank.is_complete()` and adjust the body a bit.

Reply:
Trying to think how this race condition could realistically happen. It'd require us to have a bank that is complete, and `get_consume_bank` takes so long that the new bank is already completed.

I believe the worst case of it is we process the txs and then fail to record, not great but it happens sometimes already.

I somewhat intentionally wrote this without a loop for get_consume_bank. w/o a loop I've got stronger assurance that I won't loop indefinitely for a single work item.

Reply:
got it. thanks for explaining in detail.

actually, i was nervous about whether this `assert!` is being hit or not:

https://github.com/solana-labs/solana/blob/5cd5954d21b396f5b7aec82007c927f704ca5421/runtime/src/bank.rs#L5022-L5025

however, I'm now sure that it won't happen thanks to proper `freeze_lock` in `execute_and_commit_transactions_locked()`.

seems i forgot everything the nuances of banking stage.. lol